moabb/evaluations/splitters.py

```diff
@@ -0,0 +1,126 @@
+import numpy as np
+from sklearn.model_selection import (
+    BaseCrossValidator,
+    GroupKFold,
+    LeaveOneGroupOut,
+    StratifiedKFold,
+)
+
+
+class WithinSubjectSplitter(BaseCrossValidator):
+
+    def __init__(self, n_folds):
+        self.n_folds = n_folds
+
+    def get_n_splits(self, metadata):
+        sessions_subjects = len(metadata.groupby(["subject", "session"]).first())
+        return self.n_folds * sessions_subjects
+
+    def split(self, X, y, metadata, **kwargs):
+
+        subjects = metadata.subject.values
+
+        split = IndividualWithinSubjectSplitter(self.n_folds)
+
+        for subject in np.unique(subjects):
+
+            X_, y_, meta_ = (
+                X[subjects == subject],
+                y[subjects == subject],
+                metadata[subjects == subject],
+            )
+
+            yield split.split(X_, y_, meta_)
+
+
+class IndividualWithinSubjectSplitter(BaseCrossValidator):
+
+    def __init__(self, n_folds):
+        self.n_folds = n_folds
+
+    def get_n_splits(self, metadata):
+        return self.n_folds
+
+    def split(self, X, y, metadata, **kwargs):
+
+        sessions = metadata.subject.values
+
+        cv = StratifiedKFold(self.n_folds, **kwargs)
+
+        for session in np.unique(sessions):
+            X_, y_, meta_ = (
+                X[sessions == session],
+                y[sessions == session],
+                metadata[sessions == session],
+            )
```

Comment:
```suggestion
            mask = sessions == session
            X_, y_, meta_ = (
                X[mask],
                y[mask],
                metadata[mask],
            )
```
Several times in you code, you implicitly compute the same mask multiple times. It would be more efficient (and could catch some mistakes) to compute it once only